No type inference from callback in mapped type

**TypeScript Version:**  2.7.0-dev.20180118

**Code**

```ts
declare function conforms<T>(source: { [K in keyof T]: (val: T[K]) => boolean }): (value: T) => boolean;
conforms({ foo: (v: string) => false })({ foo: "foo" });
```

**Expected behavior:**

No error. `T` is inferred as `{ foo: string }`.

**Actual behavior:**

`T` is inferred as `{ foo: {} }`.

```
src/a.ts(2,10): error TS2345: Argument of type '{ foo: (v: string) => boolean; }' is not assignable to parameter of type '{ foo: (val: {}) => boolean; }'.
  Types of property 'foo' are incompatible.
    Type '(v: string) => boolean' is not assignable to type '(val: {}) => boolean'.
      Types of parameters 'v' and 'val' are incompatible.
        Type '{}' is not assignable to type 'string'.
```

Worked in 2.7.0-dev.20180116.

Detected in `lodash` on DefinitelyTyped.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -11029,7 +11029,9 @@ namespace ts {
             const templateType = getTemplateTypeFromMappedType(target);
             const inference = createInferenceInfo(typeParameter);
             inferTypes([inference], sourceType, templateType);
-            return inference.candidates ? getUnionType(inference.candidates, UnionReduction.Subtype) : emptyObjectType;
+            return inference.candidates ? getUnionType(inference.candidates, UnionReduction.Subtype) :
+                inference.contraCandidates ? getCommonSubtype(inference.contraCandidates) :
+                emptyObjectType;
         }
 
         function getUnmatchedProperty(source: Type, target: Type, requireOptionalProperties: boolean) {

--- a/tests/baselines/reference/reverseMappedContravariantInference.js
+++ b/tests/baselines/reference/reverseMappedContravariantInference.js
@@ -1,0 +1,11 @@
+//// [reverseMappedContravariantInference.ts]
+// Repro from #21273
+
+declare function conforms<T>(source: { [K in keyof T]: (val: T[K]) => boolean }): (value: T) => boolean;
+conforms({ foo: (v: string) => false })({ foo: "hello" });
+
+
+//// [reverseMappedContravariantInference.js]
+"use strict";
+// Repro from #21273
+conforms({ foo: function (v) { return false; } })({ foo: "hello" });

--- a/tests/baselines/reference/reverseMappedContravariantInference.symbols
+++ b/tests/baselines/reference/reverseMappedContravariantInference.symbols
@@ -1,0 +1,21 @@
+=== tests/cases/compiler/reverseMappedContravariantInference.ts ===
+// Repro from #21273
+
+declare function conforms<T>(source: { [K in keyof T]: (val: T[K]) => boolean }): (value: T) => boolean;
+>conforms : Symbol(conforms, Decl(reverseMappedContravariantInference.ts, 0, 0))
+>T : Symbol(T, Decl(reverseMappedContravariantInference.ts, 2, 26))
+>source : Symbol(source, Decl(reverseMappedContravariantInference.ts, 2, 29))
+>K : Symbol(K, Decl(reverseMappedContravariantInference.ts, 2, 40))
+>T : Symbol(T, Decl(reverseMappedContravariantInference.ts, 2, 26))
+>val : Symbol(val, Decl(reverseMappedContravariantInference.ts, 2, 56))
+>T : Symbol(T, Decl(reverseMappedContravariantInference.ts, 2, 26))
+>K : Symbol(K, Decl(reverseMappedContravariantInference.ts, 2, 40))
+>value : Symbol(value, Decl(reverseMappedContravariantInference.ts, 2, 83))
+>T : Symbol(T, Decl(reverseMappedContravariantInference.ts, 2, 26))
+
+conforms({ foo: (v: string) => false })({ foo: "hello" });
+>conforms : Symbol(conforms, Decl(reverseMappedContravariantInference.ts, 0, 0))
+>foo : Symbol(foo, Decl(reverseMappedContravariantInference.ts, 3, 10))
+>v : Symbol(v, Decl(reverseMappedContravariantInference.ts, 3, 17))
+>foo : Symbol(foo, Decl(reverseMappedContravariantInference.ts, 3, 41))
+

--- a/tests/baselines/reference/reverseMappedContravariantInference.types
+++ b/tests/baselines/reference/reverseMappedContravariantInference.types
@@ -1,0 +1,28 @@
+=== tests/cases/compiler/reverseMappedContravariantInference.ts ===
+// Repro from #21273
+
+declare function conforms<T>(source: { [K in keyof T]: (val: T[K]) => boolean }): (value: T) => boolean;
+>conforms : <T>(source: { [K in keyof T]: (val: T[K]) => boolean; }) => (value: T) => boolean
+>T : T
+>source : { [K in keyof T]: (val: T[K]) => boolean; }
+>K : K
+>T : T
+>val : T[K]
+>T : T
+>K : K
+>value : T
+>T : T
+
+conforms({ foo: (v: string) => false })({ foo: "hello" });
+>conforms({ foo: (v: string) => false })({ foo: "hello" }) : boolean
+>conforms({ foo: (v: string) => false }) : (value: { foo: any; }) => boolean
+>conforms : <T>(source: { [K in keyof T]: (val: T[K]) => boolean; }) => (value: T) => boolean
+>{ foo: (v: string) => false } : { foo: (v: string) => boolean; }
+>foo : (v: string) => boolean
+>(v: string) => false : (v: string) => boolean
+>v : string
+>false : false
+>{ foo: "hello" } : { foo: string; }
+>foo : string
+>"hello" : "hello"
+

--- a/tests/cases/compiler/reverseMappedContravariantInference.ts
+++ b/tests/cases/compiler/reverseMappedContravariantInference.ts
@@ -1,0 +1,6 @@
+// @strict: true
+
+// Repro from #21273
+
+declare function conforms<T>(source: { [K in keyof T]: (val: T[K]) => boolean }): (value: T) => boolean;
+conforms({ foo: (v: string) => false })({ foo: "hello" });

